// @ladle/react/typings-for-build/shared/types.js
export var ModeState;
(function (ModeState) {
    ModeState["Full"] = "full";
    ModeState["Preview"] = "preview";
})(ModeState || (ModeState = {}));
export var ThemeState;
(function (ThemeState) {
    ThemeState["Light"] = "light";
    ThemeState["Dark"] = "dark";
    ThemeState["Auto"] = "auto";
})(ThemeState || (ThemeState = {}));
export var ControlType;
(function (ControlType) {
    ControlType["Boolean"] = "boolean";
    ControlType["String"] = "string";
    ControlType["Number"] = "number";
    ControlType["Complex"] = "complex";
    ControlType["Function"] = "function";
    ControlType["Radio"] = "radio";
    ControlType["InlineRadio"] = "inline-radio";
    ControlType["Select"] = "select";
    ControlType["MultiSelect"] = "multi-select";
    ControlType["Check"] = "check";
    ControlType["InlineCheck"] = "inline-check";
    ControlType["Action"] = "action";
    ControlType["Range"] = "range";
    ControlType["Background"] = "background";
})(ControlType || (ControlType = {}));
export var ActionType;
(function (ActionType) {
    ActionType["UpdateAll"] = "update-all";
    ActionType["UpdateMode"] = "update-mode";
    ActionType["UpdateAction"] = "update-action";
    ActionType["UpdateRtl"] = "update-rtl";
    ActionType["UpdateSource"] = "update-source";
    ActionType["UpdateStory"] = "update-story";
    ActionType["UpdateTheme"] = "update-theme";
    ActionType["UpdateWidth"] = "update-width";
    ActionType["UpdateControl"] = "update-control";
    ActionType["UpdateControlIntialized"] = "update-control-initialized";
    ActionType["UpdateHotkeys"] = "update-hotkeys";
})(ActionType || (ActionType = {}));
