// @ladle/react/typings-for-build/app/exports.js
import { useLadleContext } from "./src/context";
import { ActionType, ThemeState, ModeState, } from "../shared/types";
import * as msw from "msw";
export { msw };
export { useMDXComponents } from "@mdx-js/react";
export const Story = (props) => props.children;
export const Meta = (props) => props.children;
export const Description = (props) => props.children;
export { useLadleContext, ActionType, ThemeState, ModeState };
// deprecated, linkTo is just easier to use
export const useLink = () => {
    const { dispatch } = useLadleContext();
    return (value) => dispatch({ type: ActionType.UpdateStory, value });
};
export const linkTo = (value) => {
    const dispatch = window
        .ladleDispatch;
    return () => dispatch({ type: ActionType.UpdateStory, value });
};
export const action = (name) => {
    const dispatch = window
        .ladleDispatch;
    return (event = undefined) => dispatch({
        type: ActionType.UpdateAction,
        value: { name, event },
        clear: false,
    });
};
