// @ladle/react/typings-for-build/app/src/addons/ladle.js
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
import * as React from "react";
import { Ladle } from "../icons";
import { Modal, Code } from "../ui";
import config from "../get-config";
const hotkeyLegends = {
    fullscreen: "Toggle fullscreen mode",
    search: "Focus search input in the sidebar",
    nextStory: "Go to the next story",
    previousStory: "Go to the previous story",
    nextComponent: "Go to the next component",
    previousComponent: "Go to the previous component",
    control: "Toggle controls addon",
    darkMode: "Toggle dark mode",
    width: "Toggle width addon",
    rtl: "Toggle right-to-left mode",
    a11y: "Toggle accessibility addon",
    source: "Toggle story source addon",
};
const FormattedHotkey = ({ children }) => {
    if (navigator.platform.toLowerCase().includes("mac")) {
        children = children.replace(/alt/g, "⌥ opt").replace(/meta/g, "⌘ cmd");
    }
    else if (navigator.platform.toLowerCase().includes("win")) {
        children = children.replace(/meta/g, "⊞ win");
    }
    children = children.replace(/shift/g, "⇧ shift");
    children = children
        .replace(/arrowright/g, "→")
        .replace(/arrowleft/g, "←")
        .replace(/arrowup/g, "↑")
        .replace(/arrowdown/g, "↓")
        .replace(/\+/g, " ＋ ");
    return _jsx(Code, { children: children });
};
export const Button = ({ globalState }) => {
    const [open, setOpen] = React.useState(false);
    const text = "Get more information about Ladle.";
    return (_jsx("li", { children: _jsxs("button", { "aria-label": text, title: text, onClick: () => setOpen(true), className: open ? "ladle-active" : "", type: "button", children: [_jsx(Ladle, {}), _jsx("span", { className: "ladle-addon-tooltip", children: text }), _jsx("label", { children: "About Ladle" }), _jsxs(Modal, { isOpen: open, close: () => setOpen(false), label: "Dialog with information about Ladle.", children: [_jsx("h3", { children: "Hotkeys" }), globalState.hotkeys ? (_jsxs(_Fragment, { children: [_jsx("ul", { style: { listStyle: "none", marginLeft: 0, paddingLeft: 0 }, children: Object.keys(config.hotkeys).map((action) => config.hotkeys[action].length ? (_jsxs("li", { children: [_jsx("span", { style: {
                                                    display: "inline-block",
                                                    width: "200px",
                                                }, children: config.hotkeys[action].map((hotkey, i) => (_jsxs("span", { children: [_jsx(FormattedHotkey, { children: hotkey }), config.hotkeys[action].length > i + 1
                                                            ? " or "
                                                            : ""] }, hotkey))) }), _jsx("span", { style: { display: "inline-block" }, children: hotkeyLegends[action] })] }, action)) : null) }), _jsxs("p", { children: ["Hotkeys can be disabled through", " ", _jsx(Code, { children: `Story.meta = { hotkeys: false }` }), "."] })] })) : (_jsxs("p", { children: ["Hotkeys are disabled for this story by", " ", _jsx(Code, { children: "meta.hotkeys = false" }), "."] })), _jsxs("p", { children: ["Ladle is a modern and fast playground for React components powered by Vite. For more information visit", " ", _jsx("a", { href: "https://www.ladle.dev/", children: "ladle.dev" }), " or our", " ", _jsx("a", { href: "https://discord.gg/H6FSHjyW7e", children: "discord" }), "."] })] })] }) }));
};
