// @ladle/react/typings-for-build/app/src/ui.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { DialogOverlay, DialogContent } from "./dialog";
import { Close } from "./icons";
export const Button = ({ children, onClick, style, ...rest }) => {
    return (_jsx("button", { className: "ladle-button", onClick: onClick, style: style, "aria-label": rest["aria-label"], type: "button", children: children }));
};
export const Link = ({ children, href, style, }) => {
    return (_jsx("a", { className: "ladle-link", href: href, style: style, children: children }));
};
export const Code = ({ children }) => {
    return _jsx("code", { className: "ladle-code", children: children });
};
export const Modal = ({ children, close, isOpen, label, maxWidth = "40em", }) => {
    return (
    //@ts-ignore
    _jsx(DialogOverlay, { isOpen: isOpen, onDismiss: () => close(), "data-testid": "ladle-dialog-overlay", children: _jsxs(DialogContent, { "aria-label": label || "Modal", "data-testid": "ladle-dialog", style: { maxWidth }, children: [_jsx("div", { style: {
                        position: "absolute",
                        insetInlineEnd: "-6px",
                        top: "0px",
                    }, children: _jsx(Button, { onClick: () => close(), "aria-label": "Close modal", style: {
                            height: "36px",
                            width: "36px",
                            borderColor: "transparent",
                            boxShadow: "none",
                        }, children: _jsx(Close, {}) }) }), _jsx("div", { className: "ladle-addon-modal-body", children: children })] }) }));
};
